lib/widgets/autocomplete.dart

```diff
@@ -0,0 +1,174 @@
+import 'package:flutter/material.dart';
+
+import 'store.dart';
+import '../model/autocomplete.dart';
+import '../model/compose.dart';
+import '../model/narrow.dart';
+import 'compose_box.dart';
+
+class ComposeAutocomplete extends StatefulWidget {
+  const ComposeAutocomplete({
+    super.key,
+    required this.narrow,
+    required this.controller,
+    required this.focusNode,
+    required this.fieldViewBuilder,
+  });
+
+  /// The message list's narrow.
+  final Narrow narrow;
+
+  final ComposeContentController controller;
+  final FocusNode focusNode;
+  final WidgetBuilder fieldViewBuilder;
+
+  @override
+  State<ComposeAutocomplete> createState() => _ComposeAutocompleteState();
+}
+
+class _ComposeAutocompleteState extends State<ComposeAutocomplete> {
+  MentionAutocompleteView? _viewModel; // TODO different autocomplete view types
+
+  void _composeContentChanged() {
+    final newAutocompleteIntent = widget.controller.autocompleteIntent();
+    if (newAutocompleteIntent != null) {
+      final store = PerAccountStoreWidget.of(context);
+      _viewModel ??= (MentionAutocompleteView.init(store: store, narrow: widget.narrow)
+        ..addListener(_viewModelChanged));
```

Comment:
With that pair of parens, I was hoping to make it clear that the listener is only added once; that it doesn't behave like

```dart
      (_viewModel ??= MentionAutocompleteView.init(store: store, narrow: widget.narrow))
        ..addListener(_viewModelChanged);
```

which I think would add a new listener on every keystroke (!).

Reply:
I see.

I think the way to parse this statement as a reader is to treat `??=` as if it were `=` — all the assignment operators [have the same precedence](https://dart.dev/language/operators) (which in fact is the lowest, or loosest-binding, precedence of all operators). And then one routinely writes things like
```dart
  foo = bar()
    ..baz()
    ..quux();
```
and those mean `foo` gets the value of that whole right-hand side. So that tells you that this works the same way.